sql/planbuilder/cte.go

```diff
@@ -0,0 +1,173 @@
+package planbuilder
+
+import (
+	"fmt"
+	"strings"
+
+	ast "github.com/dolthub/vitess/go/vt/sqlparser"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+)
+
+func (b *PlanBuilder) buildWith(inScope *scope, with *ast.With) (outScope *scope) {
+	// resolveCommonTableExpressions operates on With nodes. It replaces any matching UnresolvedTable references in the
+	// tree with the subqueries defined in the CTEs.
+
+	// CTE resolution:
+	// - pre-process, get the list of CTEs
+	// - find uses of those CTEs in the regular query body
+	// - replace references to the name with the subquery body
+	// - avoid infinite recursion of CTE referencing itself
+
+	// recursive CTE (more complicated)
+	// push recursive half right, minimize recursive side
+	// create *plan.RecursiveCte node
+	// replace recursive references of cte name with *plan.RecursiveTable
+
+	outScope = inScope.replace()
+
+	for _, cte := range with.Ctes {
+		cte, ok := cte.(*ast.CommonTableExpr)
+		if !ok {
+			b.handleErr(sql.ErrUnsupportedFeature.New(fmt.Sprintf("Unsupported type of common table expression %T", cte)))
+		}
+
+		ate := cte.AliasedTableExpr
+		sq, ok := ate.Expr.(*ast.Subquery)
+		if !ok {
+			b.handleErr(sql.ErrUnsupportedFeature.New(fmt.Sprintf("Unsupported type of common table expression %T", ate.Expr)))
+		}
+
+		if with.Recursive {
+			switch n := sq.Select.(type) {
+			case *ast.Union:
+				b.buildRecursiveCte(outScope, n, ate.As.String(), columnsToStrings(cte.Columns))
+			default:
+				b.buildCte(outScope, ate, ate.As.String(), columnsToStrings(cte.Columns))
+			}
+		} else {
+			b.buildCte(outScope, ate, ate.As.String(), columnsToStrings(cte.Columns))
+		}
+	}
+	return
+}
+
+func (b *PlanBuilder) buildCte(inScope *scope, e ast.TableExpr, name string, columns []string) {
+	cteScope := b.buildDataSource(inScope, e)
+	b.renameSource(cteScope, name, columns)
+	inScope.addCte(name, cteScope)
+}
+
+func (b *PlanBuilder) buildRecursiveCte(inScope *scope, union *ast.Union, name string, columns []string) {
```

Comment:
Same admonishment about modifying the contents of params, rather than returning new values

Reply:
fixed